src/content/mods.cpp

```diff
@@ -333,51 +333,119 @@ void ModConfiguration::checkConflictsAndDeps()
 	resolveDependencies();
 }
 
-void ModConfiguration::resolveDependencies()
+struct PrioritySortedMod
 {
-	// Step 1: Compile a list of the mod names we're working with
-	std::set<std::string> modnames;
-	for (const ModSpec &mod : m_unsatisfied_mods) {
-		modnames.insert(mod.name);
+	PrioritySortedMod(ModSpec *mod) : spec(mod) {}
+
+	static bool sorter(const PrioritySortedMod *a, const PrioritySortedMod *b)
+	{
+		// Return true means: move 'a' towards begin() in std::sort
+		return a->deps_chain.size() > b->deps_chain.size();
 	}
 
-	// Step 2: get dependencies (including optional dependencies)
-	// of each mod, split mods into satisfied and unsatisfied
-	std::list<ModSpec> satisfied;
-	std::list<ModSpec> unsatisfied;
-	for (ModSpec mod : m_unsatisfied_mods) {
-		mod.unsatisfied_depends = mod.depends;
-		// check which optional dependencies actually exist
-		for (const std::string &optdep : mod.optdepends) {
-			if (modnames.count(optdep) != 0)
-				mod.unsatisfied_depends.insert(optdep);
-		}
-		// if a mod has no depends it is initially satisfied
-		if (mod.unsatisfied_depends.empty())
-			satisfied.push_back(mod);
-		else
-			unsatisfied.push_back(mod);
+	void addDependingMod(const std::string &modname,
+			std::unordered_map<std::string, PrioritySortedMod> &mods_by_name);
+
+	std::unordered_set<std::string> deps_chain;
+	ModSpec *spec;
+};
+
+void PrioritySortedMod::addDependingMod(const std::string &modname,
+		std::unordered_map<std::string, PrioritySortedMod> &mods_by_name)
+{
+	if (deps_chain.find(modname) != deps_chain.end()) {
+		// clang-format off
+		if (modname == spec->name)
+			warningstream << "PrioritySortedMod::addDependingMod(): Detected" <<
+					" circular dependencies in mod '" << modname << "'" << std::endl;
```

Comment:
It might be good to stop the entire mod loading process if there are circular dependencies. Else it might be that a mod is loaded but not the mod it depends on.
Example:
A depends on B, B depends on A. Both mods do exist, ergo they will both be loaded as they are not removed in step 1. In step 2 `addDependingMod` is used. A and B each have only one mod in their `deps_chain` set (themselves). Hence one of them is loaded before the other. This mustn't happen.